### PR TITLE
3 image index

### LINKED ISSUE #3: Image Index
#### As a user I want the homepage to display all saved images.

__Story__:

Now that we are saving images for our users, we want them to be able to view
the list of images in the system. You might be thinking, "what if my users only
want to view images of cats", or "what if my users only want to see the images
they uploaded". Those are great questions. Maybe you will have the opportunity
to add those features in a future story. For now, we will keep it simple and
quickly deliver a little more value to our users.

__Acceptance criteria__:
- [x] New images that are added show up on the homepage.
- [x] These images are persisted if the browser is closed or even if the
  server is restarted.
- [x] Images are not displayed wider than 400px.
- [x] Newest images appear first.

__Dependencies__:
- Save Image Link

### PR DESCRIPTION
**Pull request by dshen94**
Closes #3 

A small PR modifying the home page to display the 20 most recently added images newest first, with the its title, which links to the show page of the image, and uploader.

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -1,5 +1,7 @@
 class ApplicationController < ActionController::Base
   protect_from_forgery with: :exception
 
-  def home; end
+  def home
+    @images = Image.all.order(created_at: :desc).limit(20)
```
Comment (almgong) [inline app/controllers/application_controller.rb]: What happens if there are more than 20 images saved on the site? How/should users see them?

File: test/controllers/application_controller_test.rb [hunk 2]
```diff
@@ -1,8 +1,17 @@
...
 class StaticPagesControllerTest < ActionDispatch::IntegrationTest
+  setup do
+    @image = Image.create!(title: 'foo', url: 'https://i.redd.it/qesc0594n7921.jpg', uploader: 'bar')
+  end
+
   test 'should_get_home' do
     get root_path
     assert_response :success
   end
+
```
Comment (almgong) [inline test/controllers/application_controller_test.rb]: In the previous commit you had added a limit of 20, is it significant to test?

File: test/controllers/application_controller_test.rb [hunk 3]
```diff
@@ -1,8 +1,17 @@
 require 'test_helper'
 
 class StaticPagesControllerTest < ActionDispatch::IntegrationTest
+  setup do
```
Comment (almgong) [inline test/controllers/application_controller_test.rb]: Setup/teardown methods apply to all tests in the file. Nothing in need of change, but wanted to reinforce this since I noticed you added a setup that applies to just one test currently

File: test/controllers/application_controller_test.rb (hunk 2 above)
Comment (almgong) [reply test/controllers/application_controller_test.rb]: Is it important to test other edge cases, such as when there are no images?

File: test/controllers/application_controller_test.rb [hunk 4]
```diff
@@ -1,8 +1,17 @@
...
+  end
+
   test 'should_get_home' do
     get root_path
     assert_response :success
   end
+
+  test 'should display image' do
+    get root_path
+    assert_select '.img-thumbnail'
```
Comment (almgong) [inline test/controllers/application_controller_test.rb]: Should we test more than if there is at least one element with class `.img-thumbnail`? What other constraints are applied to the images rendered on the page?

File: app/controllers/application_controller.rb (hunk 1 above)
Comment (dshen94) [reply app/controllers/application_controller.rb]: I was thinking of eventually setting up a full index page that is separate from the home page, especially with the tag filtering functionality I will be adding later. Without a page buffer, I feel it's dangerous and unscalable at this point loading potentially thousands images on the landing page. Is this a valid concern in your opinion?

File: test/controllers/application_controller_test.rb (hunk 2 above)
Comment (dshen94) [reply test/controllers/application_controller_test.rb]: Yes, I also think it will be valuable to add a check for the number of image rendered and the 20 image limit.

As for the other case, the main reason that I didn't add the no images test case is because I feel like it's not exactly an error case when there's no images on the site, it's just that none may have been added yet. This may change after the next issue when image seeds are used. I'm considering modifying this test then.

File: test/controllers/application_controller_test.rb (hunk 4 above)
Comment (dshen94) [reply test/controllers/application_controller_test.rb]: Yeah, I think that's a good point. I'm current researching for a way to assert that the element's src is the same as the object's url property in the database.

File: test/controllers/application_controller_test.rb (hunk 2 above)
Comment (almgong) [reply test/controllers/application_controller_test.rb]: Edge cases are not necessarily always error cases - sometimes things can be written in a way such that when there is an absence of records, things break. I don't think it would happen to your current implementation, but you can imagine that it would occur if you keyed into a hash or expected there to be an element at some index of an array.

With respect to the above, the test would be checking that things continue to work (expected status codes, etc.) even when the state of system is empty.

File: test/controllers/application_controller_test.rb (hunk 4 above)
Comment (almgong) [reply test/controllers/application_controller_test.rb]: Sounds good, what about the order of images, should we check that?

File: app/controllers/application_controller.rb (hunk 1 above)
Comment (almgong) [reply app/controllers/application_controller.rb]: It's a valid concern - I ask because there are no issues that are directed at improving scalability of the front end, so you may need to create a standalone PR (or include it in the styes story) to address it. Otherwise the changes would be out-of-scope for the other issues. It's totally up to you if you want to add this feature.
